### DISCUSSION TITLE
[JAVA] How to implement OkHTTP3 with SP-API

### DISCUSSION THREAD

**Discussion by StrangeTim** [amzn/selling-partner-api-models#3784] [General]
Hi all, the title pretty much says it all. I have built the SP-API SDK locally and, by default, the ApiClient uses an old version of okhttp. I am using okhttp3 in other places in my code and there are conflicts arising because okhttp (old) and okhttp3 (new) both register at least 1 class used by the other. **So, my question is this: Is there a way to rebuild the SDK (specifically the ApiClient) so that it uses newer versions of the okhttp service?** 
At the moment, it looks to me like the ApiClient is hard coded for old okhttp based on the ApiClient.mustache file downloaded as part of the selling-partner-api-models repo (this file is located in sellingpartner-api-aa-java > resources > swagger-codegen > templates). I'm hoping someone else here has seen this before and knows something that I don't. 

Thanks for your time.

**Comment by stefnats** [comment]
You can start with this:

Change [ApiClient.mustache](https://github.com/amzn/selling-partner-api-models/blob/main/clients/sellingpartner-api-aa-java/resources/swagger-codegen/templates/ApiClient.mustache) 
It is located in `sellingpartner-api-aa-java/resources/swagger-codegen/templates/ApiClient.mustache`

Then you should look into the file and change the namespace from `com.squareup.okhttp` to `okhttp3`
Also change the constructor of `OkHttpClient` to use `OkHttpClient.Builder`

I think you should also change `MediaType.parse` to `MediaType.get` and also the `RequestBody.create` should be different.

Hope that helps!

**Reply by StrangeTim** [reply]
I realize that I could change the mustache file but I was hoping there was an easier way. Changing the mustache file is actually more work, I think than changing the generated file because as soon as you change from `com.squarup.okhttp` to `okhttp3` a ton of other errors start cropping up due to breaking changes between versions. I can (and am) going one by one through everything but it's a lot of work. I'm hoping somone else knows something we don't that streamlines the whole process.

**Reply by stefnats** [reply]
Fair point, but I have no suggestion on how to solve. 
May I ask what your need for another version is?

**Reply by StrangeTim** [reply]
I'm using OkHTTP3 for an internal API which is also used by the program I'm building for Amazon. It's causing conflicts.
